src/dbnode/storage/index_block_test.go

```diff
@@ -594,8 +597,43 @@ func TestNamespaceIndexBlockQuery(t *testing.T) {
 		defer nowLock.Unlock()
 		return now.ToTime()
 	}
+
 	opts := DefaultTestOptions()
-	opts = opts.SetClockOptions(opts.ClockOptions().SetNowFn(nowFn))
+
+	permitOpts := opts.PermitsOptions()
```

Comment:
Hm, instead of adding this purely to an existing test. Could we refactor this test to just have the setup and teardown pieces extracted into helper methods?

i.e.
```go
func TestNamespaceIndexBlockQuery(t *testing.T) {
	ctrl := xtest.NewController(t)
	defer ctrl.Finish()

	test := newNamespaceIndexBlockQueryTest(t, ctrl)
	
	for _, testCase := range []struct {
		name							string
		requireExhaustive bool
	}{
		{"allow non-exhaustive", false},
		{"require exhaustive", true},
	} {
		t.Run(testCase.name, func(t *testing.T) {
			test.run(namespaceIndexBlockQueryTestOptions{
				requireExhaustive: testCase.requireExhaustive,
			})
		})
	}
}

func TestNamespaceIndexBlockQueryWithPermits(t *testing.T) {
	ctrl := xtest.NewController(t)
	defer ctrl.Finish()

	test := newNamespaceIndexBlockQueryTest(t, ctrl)

	// Add permits related code here
	
	for _, testCase := range []struct {
		name							string
		requireExhaustive bool
	}{
		{"allow non-exhaustive", false},
		{"require exhaustive", true},
	} {
		t.Run(testCase.name, func(t *testing.T) {
			test.run(t, namespaceIndexBlockQueryTestOptions{
				requireExhaustive: testCase.requireExhaustive,
			})
		})
	}
}

type namespaceIndexBlockQueryTest struct {
	// things that are returned during a common setup method
}

func newNamespaceIndexBlockQueryTest(t *testing.T, ctrl *gomock.Controller) namespaceIndexBlockQueryTest {
	// do things and setup common used mocks like mock blocks, segment, results, etc
	return namespaceIndexBlockQueryTest{ /* things used by test */ }
}

type namespaceIndexBlockQueryTestOptions struct {
	requireExhaustive bool
}

func (test namespaceIndexBlockQueryTest) run(t *testing.T, opts namespaceIndexBlockQueryTestOptions) {
		// only queries as much as is needed (wrt to time)
		ctx := context.NewBackground()
		q := defaultQuery
		qOpts := index.QueryOptions{
			StartInclusive: t0,
			EndExclusive:	 now.Add(time.Minute),
		}
		// Lock to prevent race given these blocks are processed concurrently.
		var resultLock sync.Mutex
		// create initial span from a mock tracer and get ctx
		mtr := mocktracer.New()
		sp := mtr.StartSpan("root")
		ctx.SetGoContext(opentracing.ContextWithSpan(stdlibctx.Background(), sp))
		mockIterActive := index.NewMockQueryIterator(ctrl)
		mockIter0 := index.NewMockQueryIterator(ctrl)
		bActive.EXPECT().QueryIter(gomock.Any(), q).Return(mockIterActive, nil)
		mockIterActive.EXPECT().Done().Return(true)
		mockIterActive.EXPECT().Close().Return(nil)
		b0.EXPECT().QueryIter(gomock.Any(), q).Return(mockIter0, nil)
		mockIter0.EXPECT().Done().Return(true)
		mockIter0.EXPECT().Close().Return(nil)
		result, err := idx.Query(ctx, q, qOpts)
		require.NoError(t, err)
		require.True(t, result.Exhaustive)
		// queries multiple blocks if needed
		qOpts = index.QueryOptions{
			StartInclusive:		t0,
			EndExclusive:			t2.Add(time.Minute),
			RequireExhaustive: test.requireExhaustive,
		}
		bActive.EXPECT().QueryIter(gomock.Any(), q).Return(mockIterActive, nil)
		mockIterActive.EXPECT().Done().Return(true)
		mockIterActive.EXPECT().Close().Return(nil)
		b0.EXPECT().QueryIter(gomock.Any(), q).Return(mockIter0, nil)
		mockIter0.EXPECT().Done().Return(true)
		mockIter0.EXPECT().Close().Return(nil)
		mockIter1 := index.NewMockQueryIterator(ctrl)
		b1.EXPECT().QueryIter(gomock.Any(), q).Return(mockIter1, nil)
		mockIter1.EXPECT().Done().Return(true)
		mockIter1.EXPECT().Close().Return(nil)
		result, err = idx.Query(ctx, q, qOpts)
		require.NoError(t, err)
		require.True(t, result.Exhaustive)
		// stops querying once a block returns non-exhaustive
		qOpts = index.QueryOptions{
			StartInclusive:		t0,
			EndExclusive:			t0.Add(time.Minute),
			RequireExhaustive: test.requireExhaustive,
			SeriesLimit:			 1,
		}
		docs := []doc.Document{
			doc.NewDocumentFromMetadata(doc.Metadata{ID: []byte("A")}),
			doc.NewDocumentFromMetadata(doc.Metadata{ID: []byte("B")}),
		}
		mockQueryWithIter(t, mockIterActive, bActive, q, qOpts, &resultLock, docs)
		mockQueryWithIter(t, mockIter0, b0, q, qOpts, &resultLock, docs)
		result, err = idx.Query(ctx, q, qOpts)
		if test.requireExhaustive {
			require.Error(t, err)
			require.False(t, xerrors.IsRetryableError(err))
		} else {
			require.NoError(t, err)
			require.False(t, result.Exhaustive)
		}
		sp.Finish()
		spans := mtr.FinishedSpans()
		require.Len(t, spans, 9)
}
```